gap/oper.gi

```diff
@@ -616,46 +616,172 @@ end);
 InstallMethod(ModularProduct, "for a digraph and digraph",
 [IsDigraph, IsDigraph],
 function(D1, D2)
-  local m, n, E1, E2, edges, next, map, u, v, w, x;
+  local edge_function;
 
   if IsMultiDigraph(D1) or IsMultiDigraph(D2) then
     ErrorNoReturn("ModularProduct does not support multidigraphs,");
   fi;
 
+  edge_function := function(u, v, m, n, map)
+    local w, x, connections;
+    connections := [];
+    for w in OutNeighbours(D1)[u] do
+      for x in OutNeighbours(D2)[v] do
+        if (u = w) = (v = x) then
+          Add(connections, map(w, x));
+        fi;
+      od;
+    od;
+    for w in OutNeighbours(DigraphDual(D1))[u] do
+      for x in OutNeighbours(DigraphDual(D2))[v] do
+        if (u = w) = (v = x) then
+          Add(connections, map(w, x));
+        fi;
+      od;
+    od;
+    return connections;
+  end;
+
+  return(DIGRAPHS_GraphProduct(D1, D2, edge_function));
+end);
+
+InstallMethod(StrongProduct, "for two symmetric digraphs",
+[IsDigraph, IsDigraph],
+function(D1, D2)
+  local edge_function;
+
+  if IsMultiDigraph(D1) or IsMultiDigraph(D2) then
+    ErrorNoReturn("StrongProduct does not support multidigraphs,");
+  elif not IsSymmetricDigraph(D1) or not IsSymmetricDigraph(D2) then
+    ErrorNoReturn("StrongProduct only supports symmetric digraphs,");
+  fi;
+
+  edge_function := function(u, v, m, n, map)
+    local w, x, connections;
+      connections := [];
+      for x in OutNeighbours(D2)[v] do
+        AddSet(connections, map(u, x));
+        for w in OutNeighbours(D1)[u] do
+          AddSet(connections, map(w, x));
+        od;
+      od;
+      for w in OutNeighbours(D1)[u] do
+        AddSet(connections, map(w, v));
+      od;
+    return connections;
+  end;
+
+  return(DIGRAPHS_GraphProduct(D1, D2, edge_function));
+end);
+
+InstallMethod(CoNormalProduct, "for two symmetric digraphs",
+[IsDigraph, IsDigraph],
+function(D1, D2)
+  local edge_function;
+
+  if IsMultiDigraph(D1) or IsMultiDigraph(D2) then
+    ErrorNoReturn("CoNormalProduct does not support multidigraphs,");
+  elif not IsSymmetricDigraph(D1) or not IsSymmetricDigraph(D2) then
+    ErrorNoReturn("CoNormalProduct only supports symmetric digraphs,");
+  fi;
+
+  edge_function := function(u, v, m, n, map)
+    local w, x, connections;
+      connections := [];
+      for w in OutNeighbours(D1)[u] do
+        for x in [1 .. n] do
+          AddSet(connections, map(w, x));
+        od;
+      od;
+      for x in OutNeighbours(D2)[v] do
+        for w in [1 .. m] do
+          AddSet(connections, map(w, x));
+        od;
+      od;
+    return connections;
+  end;
+
+  return(DIGRAPHS_GraphProduct(D1, D2, edge_function));
+end);
+
+InstallMethod(HomomorphicProduct, "for two symmetric digraphs",
+[IsDigraph, IsDigraph],
+function(D1, D2)
+  local edge_function;
+
+  if IsMultiDigraph(D1) or IsMultiDigraph(D2) then
```

Comment:
It'd be better if this 5 lines of code went into `DIGRAPHS_GraphProduct` and were written as:

~~~~
if IsMultiDigraph(D1) then
ErrorNoReturn("the 1st argument (a digraph) must not satisfy IsMultiDigraph");
elif IsMultiDigraph(D2) then
...
~~~~